text/2019-03-04-hibernate-raft.md

```diff
@@ -0,0 +1,146 @@
+   it stops ticking.
+
+    1) All followers have met commit_index == last_index_of_leader,
+    2) apply_index == last_index,
+    3) its flag named `waken` is false,
+
+2. When a follower finds all followings are satisfied when handling ticks,
+   it stops ticking.
+
+    1) term == last_log_term,
+    2) has a valid leader,
+    3) its flag named `waken` is false,
+
+3. When a peer is created, it starts ticking.
+
+4. When a peer receives a proposal (write proposal/read index), it marks flag
+   `waken` to true and starts ticking.
+
+5. When a peer receives RequestVote/PreRequestVote/MsgTimeoutNow, it marks
+   flag `waken` to true and starts ticking.
+
+6. When a follower receives leader's message, it marks `waken` to false.
+
+7. When a leader finds all nodes are actively replying messages, it marks
+   `waken` to false.
+
+When a cluster is started, because of 3, all peers are ticking. It works
+just the old way. Because of 2, ticks of followers will not stop until a
+leader is elected and at lease a new log is appended. If there is no further
+proposals, because of 1, leader will stop ticking too.
+
+When both leader and followers are hibernated, a proposal arrives. Leader will
+start ticking again because of 4. And followers will tick. Note that leader
+must tick in this case so that it can keep retry sending out entries when
+network is in bad condition. And after the proposal is committed on all peers
+and applied on leader, leader will stop ticking because of 1.
+
+If during that time, leader is out of service, client will try to redirect
+requests to followers. Because of 4, followers will start ticking. They start
+campaigns in the end. And their campaign messages will wake the whole cluster
+up because of 5. Note that in such case, new leader will not stop ticking even
+it is elected and apply the new entry because there is still one out of service
+node, which is the old leader. Rule 4 can be more aggressive like if it detects
+that it has not received messages from leader for more than election_timeout,
+it can start campaign immediately.
+
+If during that time, follower is out of service and recovered before any
+proposals, it will start ticking because of 3. To prevent it from disturbing
+the leader, PreVote needs to be enabled. And leader will eventually be woken
+up and broadcasting heartbeat until all followers have been calmed down and go
+to sleep as 7 indicates. If follower is recovered after proposals, leader has
+been woken up already, and the whole cluster will be made sure to be up to
+date.
+
+If during that time, a new node is added to the cluster, because of 1, leader
+will keep ticking until the conf change is complete and new node has caught up
+all the logs. If a node is removed, and the node is isolated, we need another
+mechanism to make it removed. This will be discussed when talking about
+`CheckPeerStaleState`.
+
+### RaftLogGc
+
+Apparently, only leader needs to tick. And when raft tick is stopped, this
```

Comment:
```suggestion
Only leader needs to tick. And when raft tick is stopped, this
```

Let's make sure the RFC isn't working on assumptions. Is it true that only the leader needs to tick or not?